File: src/main/java/cn/chahuyun/economy/entity/fish/FishRanking.java
package cn.chahuyun.economy.entity.fish;

import cn.hutool.core.date.BetweenFormatter;
import cn.hutool.core.date.DateUnit;
import cn.hutool.core.date.DateUtil;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import net.mamoe.mirai.message.data.PlainText;
import net.mamoe.mirai.message.data.SingleMessage;

import java.io.Serializable;
import java.util.Date;

/**
 * 钓鱼排行
 *
 * @author Moyuyanli
 * @date 2022/12/14 15:08
 */
@Entity(name = "FishRanking")
@Table
@Getter
@Setter
public class FishRanking implements Serializable {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private int id;
    /**
     * 钓起着qq
     */
    private long qq;
    /**
     * 名称
     */
    private String name;
    /**
     * 尺寸
     */
    private int dimensions;
    /**
     * 金额
     */
    private double money;
    /**
     * 鱼竿等级
     */
    private int fishRodLevel;
    /**
     * 钓起来的时间
     */
    private Date date;
    /**
     * 钓起来的鱼
     */
    @ManyToOne(targetEntity = Fish.class, fetch = FetchType.EAGER)
    @JoinColumn(name = "FishId")
    private Fish fish;
    /**
     * 钓起来的鱼塘
     */
    @ManyToOne(targetEntity = FishPond.class, fetch = FetchType.EAGER)
    @JoinColumn(name = "FishPondId")
    private FishPond fishPond;

    public FishRanking() {
    }

    public FishRanking(long qq, String name, int dimensions, double money, int fishRodLevel, Fish fish, FishPond fishPond) {
        this.qq = qq;
        this.name = name;
        this.dimensions = dimensions;
        this.money = money;
        this.fishRodLevel = fishRodLevel;
        this.fish = fish;
        this.fishPond = fishPond;
        this.date = new Date();
    }


    /**
     * 显示排行榜信息
     *
     * @param top 名次
     * @return 消息
     * @author Moyuyanli
     * @date 2022/12/14 16:01
     */
    public SingleMessage getInfo(int top) {
        String message =
                "top:" + (top + 1) + "\n";
        if (top == 0 || top == 1 || top == 2) {
            String s = DateUtil.formatBetween(DateUtil.between(new Date(), getDate(), DateUnit.MS), BetweenFormatter.Level.MINUTE);
            message += "霸榜时间:" + s + "\n";
        }
        message +=
                "用户:" + getName() + "(鱼竿等级:" + getFishRodLevel() + ")\n" +
                        "尺寸:" + getDimensions() + "\n" +
                        "金额:" + getMoney() + "\n" +
                        "鱼:" + getFish().getName() + "(等级:" + getFish().getLevel() + ")\n" +
                        "鱼塘:" + getFishPond().getName() + "(鱼塘等级:" + getFishPond().getPondLevel() + ")";
        return new PlainText(message);
    }

    public int getId() {
        return id;
    }

    public FishRanking setId(int id) {
        this.id = id;
        return this;
    }

    public Fish getFish() {
        return fish;
    }

    public FishRanking setFish(Fish fish) {
        this.fish = fish;
        return this;
    }

    public FishPond getFishPond() {
        return fishPond;
    }

    public FishRanking setFishPond(FishPond fishPond) {
        this.fishPond = fishPond;
        return this;
    }
}

File: src/main/kotlin/cn/chahuyun/economy/entity/fish/FishRanking.kt
package cn.chahuyun.economy.entity.fish

import cn.hutool.core.date.DateUnit
import cn.hutool.core.date.DateUtil
import jakarta.persistence.*
import net.mamoe.mirai.message.data.PlainText
import net.mamoe.mirai.message.data.SingleMessage
import java.io.Serializable
import java.util.*

/**
 * 钓鱼排行
 *
 * @author Moyuyanli
 * @date 2022/12/14 15:08
 */
@Entity(name = "FishRanking")
@Table
class FishRanking(
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    var id: Int = 0,

    /**
     * 钓起着qq
     */
    var qq: Long = 0,

    /**
     * 名称
     */
    var name: String? = null,

    /**
     * 尺寸
     */
    var dimensions: Int = 0,

    /**
     * 金额
     */
    var money: Double = 0.0,

    /**
     * 鱼竿等级
     */
    var fishRodLevel: Int = 0,

    /**
     * 钓起来的时间
     */
    var date: Date? = null,

    /**
     * 钓起来的鱼
     */
    @ManyToOne(targetEntity = Fish::class, fetch = FetchType.EAGER)
    @JoinColumn(name = "FishId")
    var fish: Fish? = null,

    /**
     * 钓起来的鱼塘
     */
    @ManyToOne(targetEntity = FishPond::class, fetch = FetchType.EAGER)
    @JoinColumn(name = "FishPondId")
    var fishPond: FishPond? = null
) : Serializable {

    constructor(qq: Long, name: String?, dimensions: Int, money: Double, fishRodLevel: Int, fish: Fish?, fishPond: FishPond?) : this(
        qq = qq,
        name = name,
        dimensions = dimensions,
        money = money,
        fishRodLevel = fishRodLevel,
        fish = fish,
        fishPond = fishPond,
        date = Date()
    )

    /**
     * 显示排行榜信息
     *
     * @param top 名次
     * @return 消息
     */
    fun getInfo(top: Int): SingleMessage {
        var message = "top:${top + 1}\n"
        val rankingDate = date
        if (top in 0..2 && rankingDate != null) {
            val s = DateUtil.formatBetween(DateUtil.between(Date(), rankingDate, DateUnit.MS), cn.hutool.core.date.BetweenFormatter.Level.MINUTE)
            message += "霸榜时间:$s\n"
        }
        message += "用户:$name(鱼竿等级:$fishRodLevel)\n" +
                "尺寸:$dimensions\n" +
                "金额:$money\n" +
                "鱼:${fish?.name}(等级:${fish?.level})\n" +
                "鱼塘:${fishPond?.name}(鱼塘等级:${fishPond?.pondLevel})"
        return PlainText(message)
    }
}
